mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs

```diff
@@ -0,0 +1,1470 @@
+use crate::z_coin::storage::walletdb::wasm::tables::{WalletDbAccountsTable, WalletDbBlocksTable,
+                                                     WalletDbReceivedNotesTable, WalletDbSaplingWitnessesTable,
+                                                     WalletDbSentNotesTable, WalletDbTransactionsTable};
+use crate::z_coin::storage::wasm::{to_spendable_note, SpendableNoteConstructor};
+use crate::z_coin::z_coin_errors::ZcoinStorageError;
+use crate::z_coin::{CheckPointBlockInfo, WalletDbShared, ZCoinBuilder, ZcoinConsensusParams};
+
+use async_trait::async_trait;
+use common::log::info;
+use ff::PrimeField;
+use mm2_core::mm_ctx::MmArc;
+use mm2_db::indexed_db::{ConstructibleDb, DbIdentifier, DbInstance, DbLocked, IndexedDb, IndexedDbBuilder,
+                         InitDbResult, MultiIndex, SharedDb};
+use mm2_err_handle::prelude::*;
+use mm2_number::num_bigint::ToBigInt;
+use mm2_number::BigInt;
+use num_traits::{FromPrimitive, ToPrimitive};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::ops::Deref;
+use zcash_client_backend::address::RecipientAddress;
+use zcash_client_backend::data_api::{PrunedBlock, ReceivedTransaction, SentTransaction};
+use zcash_client_backend::encoding::{decode_extended_full_viewing_key, decode_payment_address,
+                                     encode_extended_full_viewing_key, encode_payment_address};
+use zcash_client_backend::wallet::{AccountId, SpendableNote, WalletTx};
+use zcash_client_backend::DecryptedOutput;
+use zcash_extras::{NoteId, ShieldedOutput, WalletRead, WalletWrite};
+use zcash_primitives::block::BlockHash;
+use zcash_primitives::consensus::{BlockHeight, NetworkUpgrade, Parameters};
+use zcash_primitives::memo::{Memo, MemoBytes};
+use zcash_primitives::merkle_tree::{CommitmentTree, IncrementalWitness};
+use zcash_primitives::sapling::{Node, Nullifier, PaymentAddress};
+use zcash_primitives::transaction::components::Amount;
+use zcash_primitives::transaction::{Transaction, TxId};
+use zcash_primitives::zip32::{ExtendedFullViewingKey, ExtendedSpendingKey};
+
+const DB_NAME: &str = "wallet_db_cache";
+const DB_VERSION: u32 = 1;
+
+pub type WalletDbRes<T> = MmResult<T, ZcoinStorageError>;
+pub type WalletDbInnerLocked<'a> = DbLocked<'a, WalletDbInner>;
+
+macro_rules! num_to_bigint {
+    ($value: ident) => {
+        $value.to_bigint().ok_or_else(|| {
+            $crate::z_coin::z_coin_errors::ZcoinStorageError::CorruptedData(
+                "Number is too large to fit in a BigInt".to_string(),
+            )
+        })
+    };
+}
+
+impl<'a> WalletDbShared {
+    pub async fn new(
+        builder: ZCoinBuilder<'a>,
+        checkpoint_block: Option<CheckPointBlockInfo>,
+        z_spending_key: &ExtendedSpendingKey,
+        continue_from_prev_sync: bool,
+    ) -> MmResult<Self, ZcoinStorageError> {
```

Comment:
Can't this be combined with the sql implementation?

Reply:
moved

Reply:
We can leave it as is for now. It's just harder to maintain this way.